File: grpc/src/main/scala/grpc/lib/compiler/Compiler.scala
package grpc
package lib
package compiler

import java.io.{File, FileInputStream, InputStreamReader}
import java.nio.charset.Charset

import grpc.lib.compiler.internal.{GrpLexer, GrpParser}
import grpc.lib.compiler.phase.{Globals, Phase, Structure, Types}
import grpc.lib.exception.{ErrorsInCodeException, ParsingException}
import grpc.lib.symbol.SymbolTable
import grpc.lib.util.Logger
import grpc.lib.util.Logger.LogLevel
import org.antlr.v4.runtime.tree.{ParseTree, ParseTreeProperty, ParseTreeWalker}
import org.antlr.v4.runtime.{ANTLRInputStream, CommonTokenStream}

class Compiler(path: String, debug: Boolean) {
  private val file = new File(path)
  private val is = new FileInputStream(file)
  private val reader = new InputStreamReader(is, Charset.defaultCharset())
  private val input = new ANTLRInputStream(reader)
  private val lexer = new GrpLexer(input)
  private val tokens = new CommonTokenStream(lexer)
  private val parser = new GrpParser(tokens)
  private var totalErrors = 0

  private val symTab = new SymbolTable
  private val results = new ParseTreeProperty[UnitResult]

  if (debug) {
    Logger.setMaxLevel(LogLevel.DEBUG)
  }

  private def checkParsing() {
    if (parser.getNumberOfSyntaxErrors > 0) {
      throw new ParsingException
    }
  }

  private def checkForErrors() {
    if (totalErrors > 0) {
      throw new ErrorsInCodeException(totalErrors)
    }
  }

  private def executePhase(tree: ParseTree, phase: Phase) {
    val walker = new ParseTreeWalker

    phase.setSymbolTable(symTab)
    phase.setResults(results)
    phase.setFileName(file.getName)

    val start = System.currentTimeMillis()
    walker.walk(phase, tree)
    val end = System.currentTimeMillis()

    if (phase.errorCount() > 0) {
      totalErrors += phase.errorCount()
      phase.getErrorList foreach { e => Logger.log(e, LogLevel.ERROR) }
    }

    Logger.log(s"Phase ${phase.getClass.getName}: " +
      s"${end - start} millis".toString, LogLevel.DEBUG)
  }

  def compile() {
    val tree = parser.init()
    checkParsing()

    executePhase(tree, new Structure)
    executePhase(tree, new Globals)
    executePhase(tree, new Types)

    checkForErrors()
  }
}

File: grpc/src/main/kotlin/grpc/lib/compiler/Compiler.kt
/*
 * Copyright 2016 Simón Oroño
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package grpc.lib.compiler

import grpc.lib.compiler.phase.*
import grpc.lib.exception.ErrorsInCodeException
import grpc.lib.exception.ParsingException
import grpc.lib.symbol.SymbolTable
import org.antlr.v4.runtime.tree.ParseTreeProperty
import java.nio.file.Paths
import kotlin.reflect.KClass

class Compiler(files: Array<String>) {
    private val paths = files.map { Paths.get(it) }.toTypedArray()
    private var totalErrors = 0

    private val symbolTable = SymbolTable()
    private val results = ParseTreeProperty<UnitResult>()

    private val compUnits = files.map {
        CompilationUnit(it, symbolTable, results, paths)
    }

    /**
     * Throws a [ParsingException] if the parser found syntax errors.
     */
    private fun checkParsing() {
        val count = compUnits.map { it.getNumberOfSyntaxErrors() }.sum()
        if (count > 0) {
            throw ParsingException()
        }
    }

    /**
     * Throws an [ErrorsInCodeException] if the total errors is higher than 0.
     */
    private fun checkForErrors() {
        if (totalErrors > 0) {
            throw ErrorsInCodeException(totalErrors)
        }
    }

    fun <T : Phase> executePhaseForAll(phaseClass: KClass<T>) {
        compUnits.map {
            totalErrors += it.executePhase(phaseClass)
        }

        checkForErrors()
    }

    /**
     * Handles the compilation process.
     */
    fun compile() {
        checkParsing()

        executePhaseForAll(Imports::class)
        executePhaseForAll(Globals::class)
        executePhaseForAll(Structure::class)
        executePhaseForAll(Types::class)
    }
}
